File: src/main/java/io/BCJSON.java
package io;

import common.CommonStatic;
import common.io.assets.AssetLoader;
import common.io.assets.UpdateCheck;
import common.io.assets.UpdateCheck.Downloader;
import common.io.assets.UpdateCheck.UpdateJson;
import common.pack.Context.ErrType;
import common.util.Data;
import main.MainBCU;
import main.Opts;
import page.LoadPage;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class BCJSON {

	public static final String[] PC_LANG_CODES = { "en", "jp", "kr", "zh", "fr", "it", "es", "de", "ru" };
	public static final String[] PC_LANG_FILES = { "util.properties", "page.properties", "info.properties",
			"StageName.txt", "UnitName.txt", "UnitExplanation.txt", "EnemyName.txt", "EnemyExplanation.txt", "ComboName.txt", "RewardName.txt", "proc.json", "animation_type.json", "CatFruitExplanation.txt" };
	public static final String JAR_LINK = "https://github.com/battlecatsultimate/bcu-assets/raw/master/jar/BCU-";
	public static final String ALT_LINK = "https://gitee.com/lcy0x1/bcu-assets/raw/master/jar/BCU-";

	public static void check() {
		LoadPage.prog("checking update information");
		UpdateJson json = Data.silent(UpdateCheck::checkUpdate);

		CommonStatic.Config cfg = CommonStatic.getConfig();
		List<Downloader> assets = null, musics = null, custMusics = null, libs = null, lang;
		UpdateJson.JarJson[] jars = null;
		try {
			for (UpdateJson.AnnouncementJson announce : json.pc_announcement) {
				if (cfg.receivedAnnouncements.contains(announce.id + "_PC") || MainBCU.ver < announce.min_ver || MainBCU.ver > announce.max_ver)
					continue;

                if (Opts.warningLong(announce.title + "\n" + String.join("\n\n", announce.text), "Announcement", 700, 350))
					cfg.receivedAnnouncements.add(announce.id + "_PC");
			}
			jars = getLatestJars(json);
			libs = UpdateCheck.checkPCLibs(json);
			assets = UpdateCheck.checkAsset(json, "pc");
		} catch (Exception e) {
			Opts.pop(e.getMessage(), "FATAL ERROR");
			e.printStackTrace();
			CommonStatic.def.save(false, true);
		}

		if (json != null) {
			int count = json.music;
			if (cfg.updateOldMusic) {
				try {
					musics = UpdateCheck.checkMusic(count).get();
				} catch (Exception ignored) {
					musics = new ArrayList<>();
				}
			} else
				musics = UpdateCheck.checkNewMusic(count);

			int customCount = json.customMusic;
			custMusics = UpdateCheck.checkNewCustomMusic(customCount);
		}

		ArrayList<String> langList = new ArrayList<>();
		for (String pcLangCode : PC_LANG_CODES) {
			for (String pcLangFile : PC_LANG_FILES)
				langList.add(pcLangCode + "/" + pcLangFile);
		}

		try {
			lang = UpdateCheck.checkLang(langList.toArray(new String[0])).get();
		} catch (Exception ignored) {
			lang = new ArrayList<>();
		}

		clearList(libs, true);
		clearList(assets, true);
		clearList(musics, false);
		clearList(custMusics, false);
		clearList(lang, false);

		Downloader font = UpdateCheck.checkFont();
		if (font != null) {
			LoadPage.prog(font.desc);
			while (!CommonStatic.ctx.noticeErr(() -> font.run(LoadPage.lp::accept), ErrType.DEBUG, "failed to download"))
				if (!Opts.conf("failed to download, retry?"))
					break;
		}

		while (!Data.err(AssetLoader::merge))
			if (!Opts.conf("failed to process assets, retry?"))
				CommonStatic.def.save(false, true);

		if (jars != null && jars.length != 0) {
			StringBuilder sb = new StringBuilder("New jar " + Data.revVer(jars[0].ver) + " is out. Update notes since current jar:\n\n");
			for (UpdateJson.JarJson info : jars) {
				sb.append(Data.revVer(info.ver))
						.append(": ")
						.append(info.desc)
						.append("\n");
			}
			sb.append("\nDo you want to update jar?");
			boolean updateIt = Opts.confLong(sb.toString(), 1000, 400);

			if (updateIt) {
				UpdateJson.JarJson jar = jars[0];
				String ver = Data.revVer(jar.ver);
				File target = new File(CommonStatic.ctx.getBCUFolder(), "./BCU-" + ver + ".jar");
				File temp = new File(CommonStatic.ctx.getBCUFolder(), "./temp.temp");
				String url = JAR_LINK + ver + ".jar";
				String alt = ALT_LINK + ver + ".jar";

				Downloader down = new Downloader(target, temp, "Downloading BCU-" + ver + ".jar...", false, url, alt);

				LoadPage.prog(down.desc);

				boolean done;

				while (!(done = CommonStatic.ctx.noticeErr(() -> down.run(LoadPage.lp::accept), ErrType.DEBUG,
						"failed to download")))
					if (!Opts.conf("failed to download, retry?"))
						break;

				if (done) {
					Opts.pop("Finished downloading BCU-" + ver + ".jar. Run this jar file from now on",
							"Download finished");

					CommonStatic.def.save(false, true);
				}
			}
		}
	}

	private static boolean clearList(List<Downloader> list, boolean quit) {
		boolean load = false;
		if (list != null)
			for (Downloader d : list) {
				LoadPage.prog(d.desc);
				boolean l;
				while (!(l = CommonStatic.ctx.noticeErr(() -> d.run(LoadPage.lp::accept), ErrType.DEBUG,
						"failed to download")))
					if (!Opts.conf("failed to download, retry?"))
						if (quit)
							CommonStatic.def.save(false, true);
						else
							break;
				load |= l;
			}
		return load;
	}

	private static UpdateJson.JarJson[] getLatestJars(UpdateJson json) {
		if (json == null)
			return null;

		List<UpdateJson.JarJson> jars = new ArrayList<>();
		for (UpdateJson.JarJson jar : json.pc_update) {
			if (jar == null)
				continue;
			if (MainBCU.ver < jar.ver)
				jars.add(jar);
		}

		return jars.toArray(new UpdateJson.JarJson[0]);
	}

	private static UpdateJson.JarJson getLatestJar(UpdateJson json) {
		if (json == null)
			return null;

		for (UpdateJson.JarJson jar : json.pc_update) {
			if (jar == null)
				continue;

			if (MainBCU.ver < jar.ver)
				return jar;
		}

		return null;
	}
}

File: src/io/BCJSON.kt
package io

import common.CommonStatic
import common.util.Data
import io.WebFileIO
import main.MainBCU
import main.Opts
import org.json.JSONArray
import org.json.JSONObject
import page.LoadPage
import java.io.*
import java.net.HttpURLConnection
import java.net.URL
import java.util.*

object BCJSON : WebFileIO {
    const val WEBSITE = "https://battlecatsultimate.github.io/bcu-page"
    const val GITRES = "https://github.com/battlecatsultimate/bcu-resources/raw/master/resources/"
    var ID = 0
    var cal_ver = 0
    private const val req = WEBSITE + "/api/"
    private const val path = "./assets/"
    private const val DOWNLOAD_LIBS = true
    private val cals: Array<String?>
    fun checkDownload() {
        LoadPage.Companion.prog("check download")
        var f: File
        var data: JSONObject? = null
        try {
            data = BCJSON["getUpdate.json"]
        } catch (e: IOException) {
            e.printStackTrace()
        }
        checkAssets(data)
        if (DOWNLOAD_LIBS) checkLibs(data)
        if (data != null) {
            LoadPage.Companion.prog("check jar update...")
            val upds: JSONArray = data.getJSONArray("pc-jars")
            var lf: JSONArray? = null
            var lt: JSONArray? = upds.get(0) as JSONArray
            if (lt.getInt(2) == 1) lt = null
            for (i in 0 until upds.length()) {
                val v: JSONArray = upds.get(0) as JSONArray
                if (v.getInt(2) == 1) {
                    lf = v
                    break
                }
            }
            if (lt != null && MainBCU.ver < lt.getString(0).toInt()) {
                if (Opts.updateCheck("JAR", lt.getString(1))) {
                    val ver: Int = lt.getString(0).toInt()
                    val name = "BCU-" + Data.Companion.revVer(ver) + ".jar"
                    if (WebFileIO.Companion.download(GITRES + "jar/" + name, File("./$name"), LoadPage.Companion.lp)) CommonStatic.def.exit(false) else Opts.dloadErr(name)
                }
            }
            if (lf != null && MainBCU.ver < lf.getString(0).toInt()) {
                if (Opts.updateCheck("JAR", lf.getString(1))) {
                    val ver: Int = lf.getString(0).toInt()
                    val name = "BCU-" + Data.Companion.revVer(ver) + ".jar"
                    if (WebFileIO.Companion.download(GITRES + "jar/" + name, File("./$name"), LoadPage.Companion.lp)) CommonStatic.def.exit(false) else Opts.dloadErr(name)
                }
            }
            LoadPage.Companion.prog("check text update...")
            for (i in cals.indices) if (!File(path + cals[i]).also { f = it }.exists() && !WebFileIO.Companion.download(GITRES + cals[i], f, null)) Opts.dloadErr(cals[i])
            if (cal_ver < data.getInt("cal")) {
                if (Opts.updateCheck("text", "")) {
                    for (i in cals.indices) if (!WebFileIO.Companion.download(GITRES + cals[i], File(path + cals[i]), null)) Opts.dloadErr(cals[i])
                    cal_ver = data.getInt("cal")
                }
            }
            LoadPage.Companion.prog("check music update...")
            val music: Int = data.getInt("music")
            val mus = BooleanArray(music)
            val fs = arrayOfNulls<File>(music)
            var down = false
            for (i in 0 until music) {
                mus[i] = !File(path + "music/" + Data.Companion.trio(i) + ".ogg").also { fs[i] = it }.exists()
                down = down or mus[i]
            }
            if (down && Opts.updateCheck("music", "")) for (i in 0 until music) if (mus[i]) {
                LoadPage.Companion.prog("download musics: " + i + "/" + mus.size)
                if (!WebFileIO.Companion.download(GITRES + "music/" + Data.Companion.trio(i) + ".ogg", fs[i], LoadPage.Companion.lp)) Opts.dloadErr("music #$i")
            }
        }
        var need = ZipLib.info == null
        f = File(path + "calendar/")
        if (!f.exists().let { need = need or it; need }) f.mkdirs()
        for (i in cals.indices) need = need or !File(path + cals[i]).exists()
        if (need) {
            Opts.pop(Opts.REQITN)
            CommonStatic.def.exit(false)
        }
    }

    private fun checkAssets(lib: JSONObject?) {
        if (lib != null && lib.length() > 1) {
            val libmap: MutableMap<String, String> = TreeMap<String, String>()
            val ja: JSONArray = lib.getJSONArray("pc-assets")
            val n: Int = ja.length()
            for (i in 0 until n) {
                val ent: JSONArray = ja.getJSONArray(i)
                if (ent.getString(2).toInt() <= MainBCU.ver) libmap[ent.getString(0)] = ent.getString(1)
            }
            var libs: MutableList<String>
            libs = if (ZipLib.info != null) ZipLib.info.update(libmap.keys) else ArrayList(libmap.keys)
            var updated = false
            while (libs.size > 0) {
                val str = libs[0]
                libs.remove(str)
                val desc = libmap[str]
                libmap.remove(str)
                if (!Arrays.asList<Any>(ZipLib.LIBREQS).contains(str)) if (!Opts.conf("do you want to download lib update $str? $desc")) continue
                LoadPage.Companion.prog("downloading asset: $str.zip")
                val temp = File(path + if (ZipLib.lib == null) "assets.zip" else "temp.zip")
                val downl: Boolean = WebFileIO.Companion.download(GITRES + "assets/" + str + ".zip", temp, LoadPage.Companion.lp)
                if (downl) {
                    if (ZipLib.info == null) ZipLib.init() else ZipLib.merge(temp)
                    libs = ZipLib.info.update(libmap.keys)
                }
                updated = true
            }
            if (updated) {
                try {
                    ZipLib.lib.close()
                } catch (e: IOException) {
                    Opts.ioErr("failed to save downloads")
                    e.printStackTrace()
                }
                ZipLib.init()
            }
        }
        ZipLib.check()
    }

    private fun checkLibs(lib: JSONObject?) {
        if (lib != null && lib.length() > 1) {
            val list: MutableList<String> = ArrayList()
            val ja: JSONArray = lib.getJSONArray("pc-libs")
            for (i in 0 until ja.length()) list.add(ja.getString(i))
            val flib = File("./BCU_lib/")
            if (!flib.exists()) flib.mkdirs()
            for (fi in flib.listFiles()) list.remove(fi.name)
            for (str in list) {
                LoadPage.Companion.prog("download $str")
                val temp = File("./BCU_lib/$str")
                WebFileIO.Companion.download(GITRES + "jar/BCU_lib/" + str, temp, LoadPage.Companion.lp)
            }
        }
    }

    @Throws(IOException::class)
    private operator fun get(app: String): JSONObject {
        val url = URL(req + app)
        val conn = url.openConnection() as HttpURLConnection
        conn.connectTimeout = 5000
        conn.setRequestProperty("Content-Type", "application/json; charset=UTF-8")
        conn.doOutput = true
        conn.doInput = true
        conn.requestMethod = "GET"
        val `in` = conn.inputStream
        val isr = InputStreamReader(`in`, "UTF-8")
        val result = readAll(BufferedReader(isr))
        if (!MainBCU.WRITE) println("result: $result")
        if (!result.startsWith("{")) throw IOException(result)
        val ans = JSONObject(result)
        `in`.close()
        conn.disconnect()
        return ans
    }

    private fun process(str: String): String {
        var str = str
        str = str.replace("\\'".toRegex(), "\\\\'")
        return str
    }

    @Throws(IOException::class)
    private fun read(json: String, app: String): JSONObject {
        val url = URL(req + app)
        val conn = url.openConnection() as HttpURLConnection
        conn.connectTimeout = 5000
        conn.setRequestProperty("Content-Type", "application/json; charset=UTF-8")
        conn.doOutput = true
        conn.doInput = true
        conn.requestMethod = "GET"
        val os = conn.outputStream
        os.write(json.toByteArray(charset("UTF-8")))
        os.close()
        val `in` = conn.inputStream
        val isr = InputStreamReader(`in`, "UTF-8")
        val result = readAll(BufferedReader(isr))
        if (!MainBCU.WRITE) println("result: $result")
        if (!result.startsWith("{")) throw IOException(result)
        val ans = JSONObject(result)
        `in`.close()
        conn.disconnect()
        return ans
    }

    @Throws(IOException::class)
    private fun readAll(rd: Reader): String {
        val sb = StringBuilder()
        var cp: Int
        while (rd.read().also { cp = it } != -1) {
            sb.append(cp.toChar())
        }
        return sb.toString()
    }

    init {
        cals = arrayOfNulls(33)
        val cal = "calendar/"
        cals[0] = io.cal + "event ID.txt"
        cals[1] = io.cal + "gacha ID.txt"
        cals[2] = io.cal + "item ID.txt"
        cals[3] = io.cal + "group event.txt"
        cals[4] = io.cal + "group hour.txt"
        for (i in 0..3) {
            val lang = "lang/" + CommonStatic.Lang.Companion.LOC_CODE.get(i) + "/"
            cals[i * 7 + 5] = io.lang + "util.properties"
            cals[i * 7 + 6] = io.lang + "page.properties"
            cals[i * 7 + 7] = io.lang + "info.properties"
            cals[i * 7 + 8] = io.lang + "internet.properties"
            cals[i * 7 + 9] = io.lang + "StageName.txt"
            cals[i * 7 + 10] = io.lang + "UnitName.txt"
            cals[i * 7 + 11] = io.lang + "EnemyName.txt"
            // TODO tutorial
        }
    }
}
